### DISCUSSION TITLE
Run Toit on a Pycom Wipy board

### DISCUSSION THREAD

**Discussion by nilwes** [toitlang/toit#90] [Show and tell]
## Introduction

Pycom offers several ESP32-based boards ready for production, so it makes perfect sense to use one of these with Toit. I had a Wipy lying around so in this show-and-tell we’ll have a look at that one.  It is a relatively simple process, even though it requires you to ensure the Wipy board is in “Boot mode” prior to flashing. In addition you’ll need an FTDI USB-to-Serial adapter and some jumper cables. Equipped with these basic tools, you’ll have your Wipy board up and running Toit code in a couple of minutes. No more MicroPython!

**NOTE:** This show-and-tell is for computers running Linux and macOS.

## Setting up Toit on your computer

Here is a [guide](https://github.com/toitlang/toit) to getting Toit up and running on your computer. I’ll assume you have this set up already before continuing.

Also, you'll need the Toit CLI. Here are [instructions](https://docs.toit.io/getstarted/installation) on how to install what you need:

## Required hardware

- A Wipy board (other Pycom boards might very well work just fine, but I have not tried it myself). Buy one [here](https://pycom.io/product/wipy-3-0/).
- A USB-to-Serial FTDI adapter ([Here](https://www.sparkfun.com/products/9873) is one alternative and here is [another one](https://www.sparkfun.com/products/9717).)
- Jumper wires
- A breadboard

**IMPORTANT**: There are many FTDI breakouts and cables available out there. However, *the most important thing is that the FTDI gear you use is made for 3.3V signal level.* FTDI cables that use 5V signal level will probably fry your Wipy board. 

## Powering your Wipy

Even though it might be possible to power your Wipy board using power from the FTDI, you might end up with a [brownout](https://en.wikipedia.org/wiki/Brownout_(electricity)) if the ESP32 as some point uses too high current. Therefore I used an external power supply to power the Wipy. You can for example pull the power from a LiPo battery, or you can do what I did: Scavenge an old USB cable, strip the 5V and GND cables, solder on a couple of jumper wires, and use these for powering the Pycom board via your computer USB port. 

![IMG_5408](https://user-images.githubusercontent.com/58735688/145195853-fcad9d01-b662-4dd9-b422-82cbebb640e7.JPG)

## Hooking up your hardware

First let’s have a look at the pin-out diagram for the Wipy. 

![Screenshot 2021-12-07 at 17 09 21](https://user-images.githubusercontent.com/58735688/145197685-e7268224-f743-45aa-8051-2bdb3d0ec58d.png)

The P0 and P1 pins are used for Rx and Tx, respectively. This means that we should
1. connect the Wipy P0 pin to the Tx pin of the FTDI adapter, and 
2. the Wipy P1 pin to the Rx pin of the adapter. 

Connect your power source to the 5V pin and the ground to the GND pin. Finally, plug your FTDI adapter to your USB port on your computer.

![IMG_5410](https://user-images.githubusercontent.com/58735688/145199723-725901a9-02ca-46cf-85fb-3451d6e2953a.JPG)

## Listening to your Wipy

To get a better insight in what’s going on with the Wipy, it is helpful to monitor the serial output from the device. To be able to do this, we need to figure out which port we should use. Check your ports by typing
```sh
ls -ltr /dev
```
This will list your ports ordered from oldest to newest, so if you connect your Wipy via your FTDI adapter to the USB of your computer, the port will show up somewhere near the end of this list. On macOS the port is probably called something like `cu.usbserial-AB0JJ2KA`, and on Linux it is probably called something like `tty-USB0`. 

![Screenshot 2021-12-07 at 17 33 47](https://user-images.githubusercontent.com/58735688/145195400-b9a6f13f-e0f5-4e6a-8fae-1d3308d67e0e.png)

Now you can monitor the serial communication with the Wipy by using the Toit command
```sh
toit serial monitor
```
This command will ask which port you want to listen to. Pick the port you just identified from the list of ports.

If you can’t see any output, try pressing the reset button on the Wipy. This should generate some output from the board.

## Ensuring you have your Wipy in Boot Mode

The Wipy must be put into Boot Mode prior to every flash. In order to do this, follow these steps:
- Cut the power to the Wipy
- Connect a jumper cable from pin P2 to GND
- Power up your Wipy.

Below is an image of the Wipy with the blue Boot Mode jumper cable connecting P2 to GND:

![IMG_5409](https://user-images.githubusercontent.com/58735688/145199363-a5f08d02-2081-453e-a070-78957d507670.JPG)

To verify the Wipy is in Boot Mode:
- In a terminal window, do `toit serial monitor` and pick the corresponding port.
- Press the reset button on the Wipy, or power-cycle it.
The terminal should output something like 
```sh
rst:0x1 (POWERON_RESET),boot:0x3 (DOWNLOAD_BOOT(UART0/UART1/SDIO_REI_REO_V2))
waiting for download
```

## Making and running your first Toit program
Let’s start with the Hello World application, to ensure we have everything set up correctly. First we make it by running
```sh
make esp32 ESP32_ENTRY=/Users/your/path/to/toit/examples/hello.toit
```
**NOTE:**  You have to enter the correct path to the folder to which you pulled Toit from Github. 

Your Wipy board should be in Boot Mode, so we are ready to flash your Wipy. Do this by running the following juicy command
```sh
python /Users/your/path/to/esp-idf/components/esptool_py/esptool/esptool.py --chip esp32 --port /dev/cu.usbserial-AB0JJ2KA --baud 921600 --before default_reset --after hard_reset write_flash -z --flash_mode dio --flash_freq 40m --flash_size detect 0xd000 /Users/nils/toit/toit/build/esp32/ota_data_initial.bin 0x1000 /Users/nils/toit/toit/build/esp32/bootloader/bootloader.bin 0x10000 /Users/nils/toit/toit/build/esp32/toit.bin 0x8000 /Users/nils/toit/toit/build/esp32/partitions.bin
```
**NOTE:** You must change the path to your esp-idf folder as well as the serial port to your Wipy. 

When the flashing is complete, remove the jumper cable that goes from P2 to GND and, power-cycle the device, and do
```sh
toit serial monitor
```

Power-cycle the Wipy board again, and it should spit out something like
```sh
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:188
ho 0 tail 12 room 4
load:0x40078000,len:12180
load:0x40080400,len:2944
entry 0x400805c8
clearing RTC memory: RTC memory is in inconsistent state
[flash reg] address 0x3f42c000, size 0x00134000
Hello, World!
E (516) Toit: VM exited, going into deep sleep.
```
Note the `Hello, World!` message in the serial output. You're now ready to start playing with Toit on a Pycom Wipy board!

## Stepping it up a bit...

The `Hello, World!` program is truly amazing. But its counterpart in the world of IoT is the blinking LED. Even better, the Wipy comes equipped with a multicolor [RGB LED](https://www.digikey.com/catalog/en/partgroup/ws2812-and-ws2812b-rgb-leds/75254) powered by the WS2812 chip. Luckily, there is a Toit package for driving this LED. Check out the package [here](https://pkg.toit.io/package/github.com%2Ftoitware%2Ftoit-pixel-strip@v0.1.1).

The package contains an example, `gpy.toit` specifically made for the RGB LED on the Pycom Wipy, but the driver can of course drive any WS2812-based RGB LED strip.

Happy tinkering!

**Comment by joshk** [comment]
Thanks for the write-up!

I also came across https://www.soracom.io/blog/internet-of-microcontrollers-made-easy-with-toit-x-soracom/ which makes it look a lot easier to flash the Pycom with the Toit firmware. 

Is there any downsides with using the instructions from the Soracom blog?

**Reply by nilwes** [reply]
Hi!
I would recommend you to download [Jaguar](https://github.com/toitlang/jaguar) and start there. Hook up the FTDI adapter's RX, TX, Vin and GND to your Pycom board, make sure it is in boot mode by starting it up with P2 and GND connected, and do a 
`jag flash`
This will flash Jaguar onto your Pycom board and it will go online immediately when you restart the board. Now you can do live update of your code on the board via Wifi. It is really neat!

Read through the [Readme](https://github.com/toitlang/jaguar) for Jaguar for more detailed instructions.

**Reply by joshk** [reply]
Thanks for such a quick reply!

If I'm using the [expansion board](https://pycom.io/product/expansion-board-3-0/), can I use the USB instead of needing to do the USB-to-Serial FTDI adapter bits?

**Reply by nilwes** [reply]
I have not tried it myself, but I think it should work. Then it automatically puts the ESP32 into boot mode.

**Reply by joshk** [reply]
My [FiPy](https://pycom.io/product/fipy/) arrived today, and I've tried using jag with limited success. I've also tried the instructions in the Soracom blog post, also with limited success.

For example, this is the output from `jag flash`

```
➜  ~ jag flash
Enter WiFi network (SSID): I'm an adult
Enter WiFi password for 'I'm an adult': ******
Flashing device over serial on port '/dev/cu.usbmodemPyf9efd11' ...
esptool.py v3.0
Serial port /dev/cu.usbmodemPyf9efd11
Connecting....
Chip is ESP32-D0WDQ6 (revision 1)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: f0:08:d1:cc:dc:ac
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 921600
Changed.
Configuring flash size...
Auto-detected Flash size: 8MB
Flash params set to 0x0230
Compressed 15392 bytes to 10770...
Wrote 15392 bytes (10770 compressed) at 0x00001000 in 0.2 seconds (effective 760.6 kbit/s)...
Hash of data verified.
Compressed 3072 bytes to 119...
Wrote 3072 bytes (119 compressed) at 0x00008000 in 0.0 seconds (effective 2247.3 kbit/s)...
Hash of data verified.
Compressed 1102176 bytes to 727321...
Wrote 1102176 bytes (727321 compressed) at 0x00010000 in 12.1 seconds (effective 726.4 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
```

But `jag ping` gets the error: `Error: couldn't ping the device`

and `jag monitor` looks like this:

```
➜  ~ jag monitor
Starting serial monitor of port '/dev/cu.usbmodemPyf9efd11' ...
``` 

but nothing else shows, and no output when I reset the device.

I also got the flash process to work (limited success) via the browser (adding a device) but when it waits for it to confirm that it can access the internet I then get an error.

When I checked my Unifi network monitor I can see that a device from `toit` is on the network.

So yeah, I'm a little lost of what to try next?

**Reply by joshk** [reply]
a little extra information, this stackoverflow thread is similar to what I am seeing : https://stackoverflow.com/questions/67015569/esp32-toit-provisioning-fails

When I run `toit serial provision -p wifi.ssid=mywifi -p wifi.password=mywifipassword --baud 460800 -m esp32-4mb`

This is the output:

```
✔ /dev/cu.usbmodemPyf9efd11
Detecting chip type. This can take up to 10s.
Checking if hardware has a Toit hardware identity... Not found
Looking for existing Toit hardware identities on disk
Using Toit hardware identity found on disk
Provisioning hardware on port /dev/cu.usbmodemPyf9efd11
+--------------------------+--------------------------------------+
|         SETTING          |                VALUE                 |
+--------------------------+--------------------------------------+
| Firmware                 | v1.6.6                               |
| Hardware Model           | esp32-4mb                            |
| Hardware ID              | 96ed7d1b-cc38-4c23-bcbb-c3aaad21ffdd |
| Hardware Batch & Seq. No | 2021-10-27#1670                      |
| context                  | remote                               |
| broker.cn                | device.toit.io                       |
| broker.port              | 9426                                 |
| wifi.ssid                | mywifi                         |
| wifi.password            | mywifipassword                     |
| broker.host              | device.toit.io                       |
+--------------------------+--------------------------------------+
erasing device flash
successfully erased device flash
writing device partitions
failed to write device partitions, do you want to retry? [y/n] y
erasing device flash
failed to write device partitions, do you want to retry? [y/n] n
log file added: toit-provision.96ed7d1b-cc38-4c23-bcbb-c3aaad21ffdd.log
failed to write device partitions
hardware identity has been written to file for future use
```

and this is the output from the log file:

```
➜  ~ cat toit-provision.96ed7d1b-cc38-4c23-bcbb-c3aaad21ffdd.log
Running "/Users/josh/.cache/toit/tools/esptool_darwin_v3.0 --chip esp32 --port /dev/cu.usbmodemPyf9efd11 --baud 460800 --before default_reset --after hard_reset erase_flash"
esptool.py v3.0
Serial port /dev/cu.usbmodemPyf9efd11
Connecting...
Chip is ESP32-D0WDQ6 (revision 1)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: f0:08:d1:cc:dc:ac
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 460800
Changed.
Erasing flash (this may take a while)...
Chip erase completed successfully in 1.3s
Hard resetting via RTS pin...
Running "/Users/josh/.cache/toit/tools/esptool_darwin_v3.0 --chip esp32 --port /dev/cu.usbmodemPyf9efd11 --baud 460800 --before default_reset --after hard_reset write_flash -z --flash_mode dio --flash_freq 40m --flash_size detect 0x1000 /var/folders/q7/t_p4ccxj6zl8t6zxghzf30dw0000gn/T/1959853612/bootloader1077175130 0x8000 /var/folders/q7/t_p4ccxj6zl8t6zxghzf30dw0000gn/T/1959853612/partitions3500557734 0x9000 /var/folders/q7/t_p4ccxj6zl8t6zxghzf30dw0000gn/T/1959853612/secure3927680264 0x10000 /var/folders/q7/t_p4ccxj6zl8t6zxghzf30dw0000gn/T/1959853612/ota_02103271255"
esptool.py v3.0
Serial port /dev/cu.usbmodemPyf9efd11
Connecting........_____....._____....._____....._____....._____....._____....._____

A fatal error occurred: Failed to connect to ESP32: Timed out waiting for packet header
Running "/Users/josh/.cache/toit/tools/esptool_darwin_v3.0 --chip esp32 --port /dev/cu.usbmodemPyf9efd11 --baud 460800 --before default_reset --after hard_reset erase_flash"
esptool.py v3.0
Serial port /dev/cu.usbmodemPyf9efd11
Connecting........_____....._____....._____....._____....._____....._____....._____

A fatal error occurred: Failed to connect to ESP32: Timed out waiting for packet header
```

**Reply by floitsch** [reply]
The log is a great help.
It seems like the board hasn't connected the UART flow-control lines to the reset/boot-pin.
This means that you need to press the reset (or boot?) button at the right moment. Normally that would be when you see the `_____....._____.....__`.
Since we hide those, a bit of guessing is involved, but the easiest is to press it just after saying 'y' for a retry.

**Reply by joshk** [reply]
Hi @floitsch , thanks for all your help. Would it be possible to do a remote pairing session and I can show you what's going on?

**Reply by floitsch** [reply]
@joshk happy to help. the easiest is probably to just use one of the chats.
Either discord, or the chat on toit.io. We can then take it from there.

**Reply by joshk** [reply]
Thank you so much @floitsch for helping me! Everything is working now.

The two main issues were:
1. during the `provision` process you need to time-to-time reset the board for the process to continue to the next step successfully
2. when the process says it is `reading hardware chip information`, you can let it fail, then quit the rest of the steps. All that is needed is for you to manually claim the device using `toit device claim`

Thank you Toit team for the language and platform!

**Reply by joshk** [reply]
@floitsch I think I found out why `toit serial monitor` isn't working with the FiPy. 

Essentially I don't think it is a FiPy issue but an issue with the Expansion Board v3

https://docs.pycom.io/datasheets/expansionboards/expansion3/

Under `Differences between v2.0 and v3.0` it notes:

> The FTDI chip has been replaced with a custom programmed PIC, similar to the Pysense/Pytrack/Pyscan boards. This allows our firmware update tool to automatically put the module into bootloader mode.

Additionally this points to an interesting feature of the v3 Expansion Board in that you can control bootloader mode programmatically.

**Reply by floitsch** [reply]
Interesting.
Alternatively using `toit serial monitor -a` should also work, as it doesn't try to reset the device.

**Comment by nilwes** [comment]
But it seems like the flashing of Jaguar went well. A few things to verify:

- Double-check that you're listening to the right port when doing `jag monitor`. Unplug-plugin the USB cable and see which port that disappears/reappears by running `ls -ltr /dev`
- Make sure your Wifi is a 2G Wifi. The ESP32 does not connect to a 5G Wifi.
- Make sure you are on the same Wifi (I have 2 different Wifis at home and sometimes my laptop switches and I cannot ping my Gpy)
- Once you know that your computer and your Fipy is on the same 2G Wifi, do a `jag scan` to search for jaguar devices on your local network.
